Mathlib/Computability/DFA.lean

```diff
@@ -163,7 +174,50 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [Set.mem_singleton_iff] at ha' hc'
   substs ha' hc'
   have h := M.evalFrom_of_pow hb hb'
-  rwa [mem_accepts, evalFrom_of_append, evalFrom_of_append, h, hc]
+  rwa [mem_accepts', evalFrom_of_append, evalFrom_of_append, h, hc]
 #align DFA.pumping_lemma DFA.pumping_lemma
 
+section Maps
+
+universe u' v'
+variable {α' : Type u'} {σ' : Type v'} (f : α ≃ α') (g : σ ≃ σ') (s : σ') (a : α') (x : List α')
+
+/-- Transforms the alphabet and states of a DFA. -/
+@[simps]
+def map (f : α' → α) (g : σ → σ') (h : σ' → σ) : DFA α' σ' where
+  step s a := g (M.step (h s) (f a))
+  start := g M.start
+  accept := g '' M.accept
```

Comment:
Should this be a preimage by `h` instead? Either way, I think the docstring needs to be clearer about the semantics of this operation, since I can't tell whether preimage or image is more correct.

Reply:
I suspect the answer is that we should have (at least) two things: `map : DFA α σ -> DFA α' σ` and `reindex : DFA α σ \~= DFA α σ'

Reply:
Thanks both! Agree with the preimage. It does seem nicer not having an existential.

I take @YaelDillies' naming suggestion, except I call the former `comap`, since the mapping function goes "backwards". I also considered adding an equiv version for the alphabet but that's just `⟨comap f, comap f.symm⟩` so not very interesting.